openfaas/faas-swarm
# Replace proxy handler with provider implementation

## Description

- Update the faas-provider version
- Update the gorilla mux version to enable simpler unit tests
- Refactor the function proxy handler to use the the new proxy
implementation
- Parse the dnsrr env variable during server startup instead of each
function invocation
- Unit test the function resolver to ensure that errors are propagated 

## Motivation and Context


- [X] I have raised an issue to propose this change #38 ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))




## How Has This Been Tested?



Additional unit tests have been provided to ensure that the resolver works as expected. The Proxy handler is unit tested upstream in the `faas-provider` 

I have also manually tested this by deploying to my local docker swarm using the `deploy_stack.sh` in openfaas/faas project with two modifications

1. update the docker-compose.yaml line 17 so that it has `direct_functions: "false"`
2. update the docker-compose.yaml line 44 so that it has `image:  theaxer/faas-swarm:new-proxy`

I then went to the UI on localhost, deployed a function from the Store and verified that the function invoke worked as expected.  I then did the same form the CLI, building, deploying, and then invoking a function.

## Types of changes


Note that this removes the `x-function` header functionality that allows the user to set the function name via a header value. This has been deprecated for some time and is not implemented in the `faas-provider`

## Checklist:
